### PR TITLE
feat: Add user settings page + Password Reset Tab + Save Token to DB

### LINKED ISSUE #92: feat: Add "change password" flow to user settings
### Describe the feature you'd like to request

Currently users can only be created and the initial password that is chosen at /auth/sign-up is hashed and saved to user table. Users cannot change or reset the password.

Suggested core feature
- [ ] New page for user settings accessible via the user menu that currently only includes "sign out"
- [ ] Form to set a new password while being logged in

Users who do not know there current password, cannot change it -> need password reset flow; this would require adding transactional emails which is probably out of scope to make this easy to finish. Happy to contribute if you want to go for it though.

### Describe the solution you'd like to see

- Add update route to user trpc router, user id of signed in user is available in context (ctx.session.user.id) -> src/server/api/routers/users.ts
  - Update user table using prisma

### Additional information

The project uses:

- shadcn/ui for ui components, find them in src/components/ui
- trpc for typed APIs used by the frontend, check out how the creation of new API keys works for reference
  - src/features/publicApi/components/CreateApiKeyButton.tsx
  - src/features/publicApi/server/apiKeyRouter.ts

<sub>[LF-655](https://linear.app/langfuse/issue/LF-655/feat-add-change-password-flow-to-user-settings)</sub>

### PR DESCRIPTION
**Pull request by khareyash05**
## What does this PR do?

1. Added UI for User settings page and 
2. Created a Dialog Box to send token sent notification
3. Created the server side code for saving token generated to user's DB
4. Found out a small font size issue bug in my previous PR. solved that in the same PR
Fixes #92 
image
image

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (does not change functionality, e.g. code style improvements, linting)
- [ ] This change requires a documentation update

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

**Comment by khareyash05** [comment]
Hi @marcklingen Could you please help me with the CI issue. Its kinda tough to understand what went wrong.

**Comment by khareyash05** [comment]
Hi @marcklingen. The e2e test seems to be failing and I am not able to work around as to how can this be solved. Kindly help

File: prisma/schema.prisma [hunk 1]
```diff
@@ -64,6 +64,7 @@ model User {
     createdAt     DateTime               @default(now()) @map("created_at")
     updatedAt     DateTime               @default(now()) @updatedAt @map("updated_at")
     featureFlags  String[]               @default([]) @map("feature_flags")
+    token         String?                @unique
```
Comment (flxwu) [inline prisma/schema.prisma:67]: s/token/password_reset_token

Comment (flxwu) [reply prisma/schema.prisma:67]: also add a token_expiry_time as DATETIME

File: prisma/migrations/20231225071648_token_user/migration.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+-- AlterTable
+ALTER TABLE "users" ADD COLUMN     "token" TEXT NOT NULL DEFAULT '';
```
Comment (flxwu) [inline prisma/migrations/20231225071648_token_user/migration.sql:2]: merge these 2 migrations into 1

**Review by flxwu — CHANGES_REQUESTED** [review]
hey, left you a first small round of reviews (just looked at the first couple of files :)) before @marcklingen or someone else gets to it.

Lmk what you think, happy to explain further!

File: prisma/schema.prisma (hunk 1 above)
Comment (khareyash05) [reply prisma/schema.prisma:67]: Sure. Will do that

File: src/pages/user/[tokenId]/changePassword.tsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }),
+    reenter_new_password: z.string().min(8, {
+      message: "New Password must be at least 8 characters long",
+    }),
+  });
+  const [credentialsFormError, setCredentialsFormError] = useState<
+    string | null
+  >(null);
+
+  // Credentials
```
[the remark this answers is not in the document]
Comment (khareyash05) [reply src/pages/user/[tokenId]/changePassword.tsx:32]: Hi Will address that in [https://github.com/langfuse/langfuse/pull/772](https://github.com/langfuse/langfuse/pull/772)

File: src/pages/user/[tokenId]/changePassword.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      new_password: "",
+      reenter_new_password: "",
+    },
+  });
+  async function onCredentialsSubmit(
+    values: z.infer<typeof credentialAuthForm>,
+  ) {
+    setCredentialsFormError(null);
+    try {
+      const res = await fetch("/api/auth/changepassword", {
```
[the remark this answers is not in the document]
Comment (khareyash05) [reply src/pages/user/[tokenId]/changePassword.tsx:46]: Changed that in [https://github.com/langfuse/langfuse/pull/772](https://github.com/langfuse/langfuse/pull/772)

File: src/pages/user/userSettings.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      <Header title="User Settings" />
+      <Alert>
+        <div className=" flex items-center justify-between ">
+          <div className="text-lg font-semibold">
+            {session.data?.user?.name}
+          </div>
+          <Avatar className="h-20 w-20">
+            <AvatarImage src={session.data?.user?.image ?? undefined} />
+            <AvatarFallback className="h-20 w-20 text-2xl">
+              {session.data?.user?.name
```
[the remark this answers is not in the document]
Comment (khareyash05) [reply src/pages/user/userSettings.tsx:64]: Will do that!

File: src/pages/user/userSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+              <div className="group relative flex items-start space-x-3 py-4">
+                <div className="flex-shrink-0">
+                  <span className="inline-flex h-10 w-10 items-center justify-center rounded-lg border border-gray-200 text-gray-400 group-hover:border-indigo-600 group-hover:text-indigo-600">
+                    <item.icon className="h-6 w-6" aria-hidden="true" />
+                  </span>
+                </div>
+                <div className="min-w-0 flex-1 items-center">
+                  <div className="text-sm font-medium text-gray-900">
+                    <a
+                      onClick={() => void item.onClick()}
```
[the remark this answers is not in the document]
Comment (khareyash05) [reply src/pages/user/userSettings.tsx:116]: Missed that! Will do that!

**Comment by khareyash05** [comment]
Thanks a lot for such a detailed code review. Let me know if there is something more I can do

File: src/pages/user/userSettings.tsx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+              </div>
+              <div className="flex-shrink-0 self-center">
+                <ChevronRightIcon
+                  className="h-5 w-5 text-gray-400 group-hover:text-gray-500"
+                  aria-hidden="true"
+                />
+              </div>
+            </div>
+          </li>
+          {instructionItems.map((item, itemIdx) => (
```
[the remark this answers is not in the document]
Comment (khareyash05) [reply src/pages/user/userSettings.tsx:88]: I totally agree with you. Solved this issue by creating a `UserSettingCTA` component. Kindly let me know if further changes are required.

File: prisma/migrations/20231225071648_token_user/migration.sql (hunk 2 above)
Comment (khareyash05) [reply prisma/migrations/20231225071648_token_user/migration.sql:2]: Created a new migration. Pls review

**Comment by khareyash05** [comment]
@marcklingen Ready for review!

**Comment by felipelo** [comment]
Hi @khareyash05 , can you please review this PR, it has conflicts to resolve.

In the mean time, how user management should be handled? I had a case where a user lost its password I couldn't reset it, so I just copy from a dumb user the password. Not secure, but it is what it is :) . Removing the password would force the user to set it?

Thanks.

**Comment by khareyash05** [comment]
> Hi @khareyash05 , can you please review this PR, it has conflicts to resolve.
> 
> In the mean time, how user management should be handled? I had a case where a user lost its password I couldn't reset it, so I just copy from a dumb user the password. Not secure, but it is what it is :) . Removing the password would force the user to set it?
> 
> Thanks.
Hey @felipelo . This is an  awesome idea. We can actually enable this type of functionality on the admin side to force to create new password. 

The current functionality which I am trying to implement is a simple reset link password which can be done by each and every one . However we can actually use this functionality in admin cases. CCing @marcklingen on this.

**Comment by marcklingen** [comment]
Closing this for now as implementation is potentially insecure. Will replace email/password in the future with email link which will overcome the password reset problem

**Pull request closed without merging**